File: Algorithm/LeetCode/0000-0050/0017.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<string> letterCombinations(string digits) {
        unordered_map<char, string> m = {
            {'2', "abc"},
            {'3', "def"},
            {'4', "ghi"},
            {'5', "jkl"},
            {'6', "mno"},
            {'7', "pqrs"},
            {'8', "tuv"},
            {'9', "wxyz"}
        };
        vector<string> res[2];
        res[0].reserve(100000);
        res[1].reserve(100000);
        int ptr = 0;
        unsigned int i, j, len = digits.size(), dlen;
        for (i = 0; i < len; ++i) {
            dlen = m[digits[i]].size();
            for (j = 0; j < dlen; ++j) {
                if (i != 0) {
                    for (auto &item : res[1-ptr]) {
                        res[ptr].push_back(item + m[digits[i]][j]);
                    }
                } else {
                    res[ptr].push_back(string(1, m[digits[i]][j]));
                }
            }
            ptr = 1 - ptr;
            res[ptr].clear();
        }
        return res[1-ptr];
    }
};

int main() {
    Solution s;
    auto res = s.letterCombinations("23");
    for (auto item : res) {
        cout << item << endl;
    }
    return 0;
}

File: Algorithm/Code/LeetCode/Python/0000-0050/0017.py
#!/bin/python3
# coding: utf-8

"""
Given a digit string, return all possible letter combinations that the number could represent.

A mapping of digit to letters (just like on the telephone buttons) is given below.



Input:Digit string "23"
Output: ["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"].
Note:
Although the above answer is in lexicographical order, your answer could be in any order you want.
"""
from __future__ import print_function

__author__ = "bovenson"
__email__ = "szhkai@qq.com"
__date__ = "2017-11-08 17:31"


class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """
        d_s = {
            '2': 'abc',
            '3': 'def',
            '4': 'ghi',
            '5': 'jkl',
            '6': 'mno',
            '7': 'pqrs',
            '8': 'tuv',
            '9': 'wxyz'
        }
        _last = ['']
        _res = []
        for i in range(len(digits)):
            _res = []
            for c in d_s[digits[i]]:
                for s in _last:
                    _res.append(s + c)
            _last = _res
        return _res


if __name__ == "__main__":
    res = Solution().letterCombinations('23')
    print(res)
